### PR TITLE
High level wrapper over collaberation (for discussion)

### PR DESCRIPTION
**Pull request by samdroid-apps**
Probs requires https://github.com/sugarlabs/sugar-toolkit-gtk3/pull/270 if your a system that doesn't give out telepathy tubes

This commit introduces the CollabWrapper.  This is a high level
wrapper over the collaberation system.  It uses telepathy text
channels (rather than deprecated tubes used in most activities)
as well as creating file transfer channels to transfer large
messages in a 1-to-1 manner.

For activities, the CollabWrapper provides a simple api.  It syncs
the state when a user joins the activity.  It allows for posting
messages to all other buddies.  It also allows for initiating a
file transfer directly to a buddy, from either a string in memory
or a file on the file system.

Example API usage:  https://github.com/samdroid-apps/bibliography-activity/compare/collab?expand=1

TODO:

* Further test resuming a shared activity
* Discuss if file transfer tubes should be considered backwards
  compatiable, as they popup shell notifications without the
  appropriate patch (https://github.com/sugar/pull/621)
* Test sending a file from a file on the file system
* Test accepting a file to the filesystem

### CODE REVIEW

**Comment by samdroid-apps** [comment]
Preview of documentation on this branch:  http://people.sugarlabs.org/sam/docs-collab-wrapper-try2/

Collab wrapper class:  http://people.sugarlabs.org/sam/docs-collab-wrapper-try2/sugar3.presence.wrapper.html

File transfer abstractions:  http://people.sugarlabs.org/sam/docs-collab-wrapper-try2/sugar3.presence.filetransfer.html

Another item for the TODO is to port jarabe to use the file transfer api in this patch rather than in jarbae.model.filetransfer.  However, that is simple compared to this patch, and could eaisly be done later as both apis are very simmilar.

**Comment by samdroid-apps** [comment]
Great, ran through some more tests and resuming a saved activity works fine.  I tested the following test case:

1.  Alice opens up a previously shared saved activity
2.  Alice invites Bob
3.  Bob joins and adds new content
4.  Alice sees the new content

It doesn't work if 2 people resume the shared activity from their journals.  Is that meant to work?  I think it is probably an issue in the code that decides who will lead the activity.

**Comment by walterbender** [comment]
We should check, but I always thought that the sharer (leader) was not
something that could be handed off to other participants. There are
countless corner cases that never really got sorted out definitively.

On Sun, Dec 6, 2015 at 3:22 PM, Sam <notifications@github.com> wrote:

> Great, ran through some more tests and resuming a saved activity works
> fine. I tested the following test case:
>
>    1. Alice opens up a previously shared saved activity
>    2. Alice invites Bob
>    3. Bob joins and adds new content
>    4. Alice sees the new content
>
> It doesn't work if 2 people resume the shared activity from their
> journals. Is that meant to work? I think it is probably an issue in the
> code that decides who will lead the activity.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/sugarlabs/sugar-toolkit-gtk3/pull/272#issuecomment-162344616>
> .
>



-- 
Walter Bender
Sugar Labs
http://www.sugarlabs.org

**Comment by samdroid-apps** [comment]
oh, thats not good.

I only use the leader in this wrapper to deal with initial state sync.  However, that could probably replaced with all initized buddies offering to send the state to the noob, and the noob only accepting 1 file transfer.

But that still doesn't help with the resuming shared activity thing I don't think.  Do we need to sync state when people resume a shared activity?  I'm not sure how to implement it if we do because how do we know if the buddy is the first to come to the shared activity (and should be marked initized) or is joining a shared activity by resuming it?????  I'm very confused right now/

**Comment by walterbender** [comment]
Don't be confused. Some of the corner cases -- which are largely
activity-specific -- were not ever resolved.

On Sun, Dec 6, 2015 at 3:45 PM, Sam <notifications@github.com> wrote:

> oh, thats not good.
>
> I only use the leader in this wrapper to deal with initial state sync.
> However, that could probably replaced with all initized buddies offering to
> send the state to the noob, and the noob only accepting 1 file transfer.
>
> But that still doesn't help with the resuming shared activity thing I
> don't think. Do we need to sync state when people resume a shared activity?
> I'm not sure how to implement it if we do because how do we know if the
> buddy is the first to come to the shared activity (and should be marked
> initized) or is joining a shared activity by resuming it????? I'm very
> confused right now/
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/sugarlabs/sugar-toolkit-gtk3/pull/272#issuecomment-162346132>
> .
>



-- 
Walter Bender
Sugar Labs
http://www.sugarlabs.org

**Comment by quozl** [comment]
Typo in commit description first line.

**Comment by samdroid-apps** [comment]
@quozl sorry, I don't see it.

Any other comments on this patch?

**Comment by samdroid-apps** [comment]
For reference, here is the patch for browse:  https://github.com/sugarlabs/browse-activity/pull/17

**Comment by quozl** [comment]
@samdroid-apps wrote:
> @quozl sorry, I don't see it.

It isn't how we spell collaboration.  Also in remainder of commit.  In sed terms, s/collaberation/collaboration/g

**Comment by quozl** [comment]
I'm yet to see a feature page, and once I do there'll be a ton of questions. But looking at the patch for browse makes me think "CollabWrapper" as a name won't be so easily understood.  Perhaps give it a name that is shorter, and that doesn't expose implementation detail.  Go look for synonyms of collaboration.

**Comment by samdroid-apps** [comment]
Ok, you can see a feature page:  https://wiki.sugarlabs.org/go/Features/Fixing_Collab_(Tubes)

If you have any preferences for another name feel free to suggest.

File: src/sugar3/presence/filetransfer.py [hunk 1]
```diff
@@ -0,0 +1,358 @@
...
+    def __init__(self, buddy, conn, path, filename, description, mime):
+        BaseOutgoingTransfer.__init__(
+            self, buddy, conn, filename, description, mime)
+
+        self._path = path
+        file_size = os.stat(path).st_size
+        self._create_channel(file_size)
+
+    def _get_input_stream(self):
+        logging.debug('opening %s for reading', self._file_name)
```
Comment (samdroid-apps) [inline src/sugar3/presence/filetransfer.py]: Oops, this is an issue.  I'll fix this up an prepare a patch for image viewer

**Comment by samdroid-apps** [comment]
Development moved to https://github.com/samdroid-apps/collabwrapper

Maybe I'll try to merge this next cycle

**Comment by nullr0ute** [comment]
Can we call it sugar-collabwrapper if it needs to be packaged separately, I'd sooner it be part of toolkit or the core sugar though. Does it mean it's not being fixed the 107 cycle (again)?

**Comment by samdroid-apps** [comment]
No more packages will be requires, as the code in included in activities git repos (sometimes as submodules, other times just copied and pasted).

The new api for activities will not be included in sugar3 during the 10{7,8} cycle.  However some activities have began using it by including a copy of the  https://github.com/samdroid-apps/collabwrapper repository, which will allow for faster (no timeline) development between now and the next cycle.
